src/mlpack/methods/ann/layer/minibatch_discrimination_impl.hpp

```diff
@@ -0,0 +1,143 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MINIBATCH_DISCRIMINATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MINIBATCH_DISCRIMINATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "minibatch_discrimination.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+MiniBatchDiscrimination<InputDataType, OutputDataType
+>::MiniBatchDiscrimination() :
+  A(0),
+  B(0),
+  C(0),
+  batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+MiniBatchDiscrimination<InputDataType, OutputDataType
+>::MiniBatchDiscrimination(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t features) :
+    A(inSize),
+    B(outSize - inSize),
+    C(features),
+    batchSize(0)
+{
+  weights.set_size(A * B * C, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void MiniBatchDiscrimination<InputDataType, OutputDataType>::Reset()
+{
+  weight = arma::mat(weights.memptr(), B * C, A, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void MiniBatchDiscrimination<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  batchSize = input.n_cols;
+  tempM = weight * input;
+  M = arma::cube(tempM.memptr(), B, C, batchSize, false, false);
+  distances.set_size(B, batchSize, batchSize);
+  output.set_size(B, batchSize);
+  for (size_t i = 0; i < M.n_slices; i++)
+  {
+    output.col(i).ones();
+    for (size_t j = 0; j < M.n_slices; j++)
+    {
+      if (j < i)
+      {
+        output.col(i) += distances.slice(j).col(i);
+      }
+      else if (i == j)
+      {
+        continue;
+      }
+      else
+      {
+        distances.slice(i).col(j) =
+          arma::exp(-arma::sum(abs(M.slice(i) - M.slice(j)), 1));
+        output.col(i) += distances.slice(i).col(j);
+      }
+    }
+  }
+  output = join_cols(input, output); // (A + B) x batchSize
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void MiniBatchDiscrimination<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  g = gy.head_rows(A);
+  arma::Mat<eT> gM = gy.tail_rows(B);
+  deltaM.zeros(B, C, batchSize);
+  for (size_t i = 0; i < M.n_slices; i++)
+  {
+    for (size_t j = 0; j < M.n_slices; j++)
+    {
+      if (i == j)
+      {
+        continue;
+      }
+      arma::mat t = arma::sign(M.slice(i) - M.slice(j));
+      t.each_col() %=
+          distances.slice(std::min(i, j)).col(std::max(i, j)) % gM.col(i);
+      deltaM.slice(i) -= t;
+      deltaM.slice(j) += t;
+    }
+  }
+  deltaTemp = arma::mat(deltaM.memptr(), B * C, batchSize, false, false);
+  g += weight.t() * deltaTemp;
```

Comment:
Hmm, could you comment more on how you are calculating the derivative?

Reply:
This is just the backpropogation for this line:
https://github.com/mlpack/mlpack/pull/1913/files#diff-95a80beec0526fd4dcfb849346210aeeR58

Reply:
I meant the entire function in general?

Reply:
```
distances.slice(i).col(j) =
          arma::exp(-arma::sum(abs(M.slice(i) - M.slice(j)), 1));
```
if you differentiate this wrt to `M.slice(i)` then you would get 
distances.slice(i).col(j) * d (-arma::sum(abs(M.slice(i) - M.slice(j)), 1) / d M.slice(i)
 = distances.slice(i).col(j) * sign(M.slice(i) - M.slice(j))
and then similarly differentiating this wrt `M.slice(j)` gives the same result with opposite sign.
which leads to
```
deltaM.slice(i) -= t;
deltaM.slice(j) += t;
```
Then finally when we have computed the delta for the features (M cube) we can compute the backprop to the input. This step of matrix multiplication with the weights is analogous to that in a linear layer.

Reply:
I see, thanks for the clarification.